add deleteVertices to QgsAbstractGeometry

ViperMiniQ opened the pull request:
This PR implements `deleteVertices` for `QgsAbstractGeometry` classes and is the first part of #62669. 
Unfortunately, some types of geometries can hold other types of geometries, so the method in QgsGeometry as tried in #62669 cannot properly handle all and that is why I decided for this approach.

Follow up PR will modify the QgsVertexTool to use the newly created functions which will avoid constant calls to detach and speed up vertices deletion on complex geometries as well as simplify the vertex tool delete function.

I need to add more tests for CoumpundCurve and CurvePolygon types before setting this PR as ready. Suggestions are welcome.

Even the special case of a "middle vertex" of a circularstring as added in #53960 is checked for.

(ping @uclaros and @3nids)

ViperMiniQ committed:
add deleteVertices to geometries

ViperMiniQ committed:
add tests

ViperMiniQ committed:
run astyle

ViperMiniQ committed:
try handling cppcheck warning out of bounds

ViperMiniQ committed:
suppress containerOutOfBounds warning

ViperMiniQ committed:
solve strange behavior when deleting first/last point of compoundcurve

ViperMiniQ committed:
fix deletion of a middle vertex in compound curves which have more than one circularstring arc

ViperMiniQ committed:
fix formatting and spell check warnings

ViperMiniQ committed:
use signed int for size

ViperMiniQ committed:
add more test cases for curvepolygon which has compoundcurve ring

ViperMiniQ commented:
The purpose of this PR is to clearly define the behavior when multiple vertices are deleted from a geometry, since not all geometries can be modified as easily, there are some rules we need/want to follow.

This change also provides much better result when multiple vertices are deleted through the vertex tool. Currently, the vertex tool filters out the start and end points of polygons and removes selected points if the geometry would otherwise be deleted before reaching them.
None of that is needed as `deleteVertices` method handles all.

Another important aspect is the `detach()` call, which is now executed only once within this method. This results in a noticeable performance improvement.

To test the vertex tool modifications, you will need the following commit (you can also review the parts of the code that will become obsolete in the vertex tool, as they will no longer be needed):
https://github.com/ViperMiniQ/QGIS/commit/9380b8b9ab8284a39eede5e3798e5e84ae3d7bfc

If this PR is accepted and merged, that commit will also become a PR.

With the vertex tool updated to use the `deleteVertices` method, we get a significant performance increase when deleting geometry vertices.

------------

Testing dataset: [ne_10m_land.zip](https://github.com/user-attachments/files/23292481/ne_10m_land.zip)

Test 1:
entire map would be selected (all features, all vertices) with the vertex tool and delete key would be pressed.

Time: 
`deleteVertices` – ~15 seconds
`deleteVertex` – didn’t finish after 40 minutes, filled up the entire RAM and crashed the system (no other apps were running while the deletion was ongoing)

------------

Test 2:
roughly half of Antarctica would be selected with the vertex tool and delete key pressed

Time:
`deleteVertices` – ~1.5 seconds
`deleteVertex` -  ~25 seconds

Apart from speed, we will save enough energy to train better AIs to replace us :)

------------

Creating a geometry in Python without registering it (this is important because this just modifies the geometry in place, so functions should have similar times).

`QgsGeometry.fromWkt()` and iterating `deleteVertex` vs a single call to `deleteVertices` (same dataset as above):

No of vertices deleted | Time deleteVertex (s) | Time deleteVertices (s)
-- | -- | --
270415 | 26.589434 | 0.590351
35369 | 70.662433 | 3.793866
31693 | 17.344827 | 0.878986
26548 | 5.149592 | 0.282969
26035 | 34.360964 | 1.739295
19457 | 1.021254 | 0.065217
12458 | 4.921302 | 0.256777
11639 | 3.676333 | 0.188385
7448 | 0.438811 | 0.027785
4998 | 0.111671 | 0.009919

------------

Creating a layer and iterating its features calling `deleteVertex` and `deleteVertices` on `feature.geometry()`:

No of vertices deleted | Time deleteVertex (s) | Time deleteVertices (s)
-- | -- | --
270415 | 26.589434 | 0.590351
35369 | 70.662433 | 3.793866
31693 | 17.344827 | 0.878986
26548 | 5.149592 | 0.282969
26035 | 34.360964 | 1.739295
19457 | 1.021254 | 0.065217
12458 | 4.921302 | 0.256777
11639 | 3.676333 | 0.188385
7448 | 0.438811 | 0.027785
4998 | 0.111671 | 0.009919

------------

Right now, using the vertex tool is best being done on a single vertex, otherwise the tool might end up deleting more vertices than selected, there is no support for multiple vertices deletion on CompoundCurve geometries, for example (till now).

Here is how the tool behaves on master vs the behavior with deleteVertices:

[Screencast_20251102_203328.webm](https://github.com/user-attachments/assets/82cb8290-9e62-40c4-a07d-ffc2118da43e)
What is that? Notice how the first vertex behaves vs all the others, how many more vertices are being deleted than are being selected and how I am unable to delete vertices if I select the entire bottom part or if I select the entire left part.

[Screencast_20251102_202711.webm](https://github.com/user-attachments/assets/b38a1aa5-f4c7-4dae-b621-6fe14bf2d61d)
Here, for all mentioned cases, we end up with a much more expected result.

That weird geometry if you want to play with it:
`CurvePolygon (CompoundCurve (CircularString (3 19, 0 22, 3 25, 6 28, 9 25),( 9 25, 14 20, 19 25 ),CircularString (19 25, 22 28, 25 25, 28 22, 25 19),(25 19, 20 14, 25 9),CircularString (25 9, 28 6, 25 3, 22 0, 19 3),(19 3, 14 8, 9 3),CircularString (9 3, 6 0, 3 3, 0 6, 3 9),(3 9, 8 14, 3 19) ))))`

:)

uclaros commented on src/core/geometry/qgslinestring.cpp:
```diff
@@ -2088,6 +2088,43 @@ bool QgsLineString::deleteVertex( QgsVertexId position )
+    if ( position.vertex >= mX.size() || position.vertex < 0 )
+    {
+      return false;
+    }
```
Since positions is sorted, we can check `positions.constFirst().vertex >= mX.size()` and `positions.constLast().vertex < 0` once before starting the iteration.

uclaros reviewed (commented):
This is a partial review, only for `QgsAbstractGeometry`, `QgsGeometry` and `QgsLinestring`.

Let's first decide on the final behavior regarding passing invalid vertex ids and what happens when a degenerate geometry is produced after a vertex deletion (fail vs clear geom).

ViperMiniQ committed:
pass by const ref, fix docstring

ViperMiniQ commented on src/core/geometry/qgsabstractgeometry.h:
```diff
@@ -527,6 +527,14 @@ class CORE_EXPORT QgsAbstractGeometry
+    /**
+     * Deletes vertices within the geometry
+     * \param positions list of vertex ids for vertices to delete
+     * \returns TRUE if delete was successful
```
A successful deletion is when all the requested vertices are actually deleted (maybe some extra as well, to keep the geometry valid), but there is a difference between `QgsAbstractGeometry` and `QgsGeometry`. 

If we reach a state where less than required vertices are present to form a valid geometry (linestring with less than 2 vertices, a polygon with less than 3), we clear the geometry at that point and return TRUE, same as in `deleteVertex`.

If any of the `QgsAbstractGeometry` vertices fails to be deleted, the method should return FALSE. This can potentially leave the geometry in an invalid state. 

Invalid `VertexIds` are not handled at all in `QgsAbstractGeometry`, the same way they are not handled in the `deleteVertex` method. Were you thinking of a check to see if the requested VertexIds actually belong to the the geometry? For example, right now, neither `deleteVertex` or `deleteVertices` care for the `part_id` or `ring_id` in case of linestring, they only use the `vertex` bit of the VertexId.

Calling `deleteVertices` on a `QgsGeometry` should always produce a valid geometry – it internally calls the `QgsAbstractGeometry::deleteVertices` and if it gets a FALSE from it, it resets the geometry to whatever it was before the call (in `QgsGeomtry::deleteVertices`, we copy the geometry so we can reset to it if the deletion of any vertices fails).

ViperMiniQ committed:
handle points deletion in geometry collection, add tests

ViperMiniQ committed:
fix docstring

ViperMiniQ committed:
sipify

ViperMiniQ committed:
sipify

ViperMiniQ commented:
Take my hand, we'll make it, I swear

ViperMiniQ commented:
Oh, we've gotta hold on, ready or not

ViperMiniQ commented:
You live for the fight when that's all that you've got

uclaros commented:
I wonder whether we should be passing a QSet of vertices to be deleted instead of QList... What do you think?

ViperMiniQ commented:
> I wonder whether we should be passing a QSet of vertices to be deleted instead of QList... What do you think?


I think I'd be a good idea. It would document itself and we don't have to worry about duplicates within geometry methods.

ViperMiniQ committed:
Merge branch 'master' into add-deletevertices

ViperMiniQ committed:
run precommit

ViperMiniQ committed:
change qlist to qset
change since 4.0 to 4.2

ViperMiniQ commented:
Changed `QgsAbstractGeometry` and `QgsGeometry` to accept `QSet` instead of `QList`.
Only the signature has changed, internally it is still copying to `QList` which is then sorted for vertex deletion, so no major changes.

@lbartoletti, would you mind taking a look at the nurbscurve implementation for this method?
TBH, I am still not overly familiar with nurbs, so I modeled it based on rough understanding of it and the logic from deleteVertex(). You also didn't supply any tests to check `deleteVertex()`, so I also left out the tests for `deleteVertices()` for nurbs.

gacarrillor commented:
> Here is how the tool behaves on master vs the behavior with deleteVertices:
> 
> [Screencast_20251102_203328.webm](https://github.com/user-attachments/assets/82cb8290-9e62-40c4-a07d-ffc2118da43e) What is that? Notice how the first vertex behaves vs all the others, how many more vertices are being deleted than are being selected and how I am unable to delete vertices if I select the entire bottom part or if I select the entire left part.
> 
> [Screencast_20251102_202711.webm](https://github.com/user-attachments/assets/b38a1aa5-f4c7-4dae-b621-6fe14bf2d61d) Here, for all mentioned cases, we end up with a much more expected result.
> 
> That weird geometry if you want to play with it: `CurvePolygon (CompoundCurve (CircularString (3 19, 0 22, 3 25, 6 28, 9 25),( 9 25, 14 20, 19 25 ),CircularString (19 25, 22 28, 25 25, 28 22, 25 19),(25 19, 20 14, 25 9),CircularString (25 9, 28 6, 25 3, 22 0, 19 3),(19 3, 14 8, 9 3),CircularString (9 3, 6 0, 3 3, 0 6, 3 9),(3 9, 8 14, 3 19) ))))`
> 
> :)

@ViperMiniQ, nice demo for the compound curve scenario.

Looks like the issue posted by @troopa81 in [this comment](https://github.com/qgis/QGIS/pull/53960#pullrequestreview-1573117479) is solved with `deleteVertices`, right?

ViperMiniQ commented:
> Looks like the issue posted by @troopa81 in [this comment](https://github.com/qgis/QGIS/pull/53960#pullrequestreview-1573117479) is solved with `deleteVertices`, right?

@gacarrillor 

[Screencast_20260523_213301.webm](https://github.com/user-attachments/assets/1dd920a8-c694-441c-a35e-55abe3d80959)

The screencasts from the comment you quoted shows exactly that, although a bit more exaggerated.

uclaros commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -759,6 +759,41 @@ bool QgsPolyhedralSurface::deleteVertex( QgsVertexId vId )
+      {
+        delete mPatches.takeAt( part );
+      }
+      clearCache();
```
Could `clearCache()` be called once outside the loop?

uclaros commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -759,6 +759,41 @@ bool QgsPolyhedralSurface::deleteVertex( QgsVertexId vId )
+
+    QgsPolygon *patch = mPatches.at( part );
+
+    if ( patch->deleteVertices( vertexMap ) )
```
If a patch fails to delete the vertices, shouldn't the whole method return `false`?

uclaros commented on src/core/geometry/qgspolyhedralsurface.cpp:
```diff
@@ -759,6 +759,41 @@ bool QgsPolyhedralSurface::deleteVertex( QgsVertexId vId )
+  QMap<int, QSet<QgsVertexId>> partVertices;
+  for ( QgsVertexId pos : positions )
+  {
+    partVertices[pos.part].insert( pos );
```
I don't think we should be ignoring the `part` part of the vertex id (it is often ignored for geoms that do not consider having multiple parts).
```suggestion
    partVertices[pos.part].insert( QgsVertexId( 0, pos.ring, pos.vertex ) );
```

ViperMiniQ replied:
hasVertex method has been introduced to check if a given QgsVertexId actually belongs to the geometry. QgsGeometryCollection strips the part bit before passing it on, QgsCurvePolygons strip the ring bit when passing it on.

uclaros commented on src/core/geometry/qgslinestring.cpp:
```diff
@@ -2083,6 +2083,46 @@ bool QgsLineString::deleteVertex( QgsVertexId position )
   return true;
 }
 
+bool QgsLineString::deleteVertices( const QSet<QgsVertexId> &vertices )
```
Other implementation use `positions` for this parameter. Let's follow the pattern here too, and use a different name for the `positions` sorted list.
```suggestion
bool QgsLineString::deleteVertices( const QSet<QgsVertexId> &positions )
```

uclaros commented on src/core/geometry/qgsgeometrycollection.cpp:
```diff
@@ -679,6 +679,49 @@ bool QgsGeometryCollection::deleteVertex( QgsVertexId position )
+    // quit if any vertex on any part fails to be deleted
+    if ( !geom->deleteVertices( partVertices ) )
+    {
+      return false;
```
Won't this leave the geometry collection in a semi-modified state?
If _some_ vertices have been deleted in previous iterations I think we should revert to the original geometry.

ViperMiniQ replied:
Hm, I was thinking about having that revert only for QgsGeometry container and not specific abstract implementations. Should the same be performed for CompoundCurve and QgsCurvePolygon?

uclaros replied:
This is resolved by checking early using the new `hasVertex()` method

uclaros commented on src/core/geometry/qgsgeometrycollection.cpp:
```diff
@@ -679,6 +679,49 @@ bool QgsGeometryCollection::deleteVertex( QgsVertexId position )
+  QMap<int, QSet<QgsVertexId>> partVertices;
+  for ( QgsVertexId pos : positions )
+  {
+    partVertices[pos.part].insert( pos );
```
I don't think we should be ignoring the part part of the vertex id (it is often ignored for geoms that do not consider having multiple parts).
```suggestion
    partVertices[pos.part].insert( !gsVertexId( 0, pos.ring, pos.vertex ) );
```

ViperMiniQ replied:
hasVertex method has been introduced to check if a given QgsVertexId actually belongs to the geometry. QgsGeometryCollection strips the part bit before passing it on, QgsCurvePolygons strip the ring bit when passing it on.

uclaros commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -667,6 +667,47 @@ bool QgsGeometry::deleteVertex( int atVertex )
+  // if it is a point, set the geometry to nullptr
+  if ( QgsWkbTypes::flatType( d->geometry->wkbType() ) == Qgis::WkbType::Point )
+  {
+    if ( atVertices.size() != 1 && !atVertices.contains( 0 ) )
+      return false;
```
This does not reflect the docstring description.
TBH, I prefer this logic over the `deleteVertex()` one that _ignores_ the passed vertex id.

uclaros reviewed (commented):
Another partial review for more abstract geometry classes.
Last ones I still haven't checked are curved geometries

lbartoletti commented:
> @lbartoletti, would you mind taking a look at the nurbscurve implementation for this method?

I'll try to this week

ViperMiniQ committed:
add hasVertex method
hasVertex checks that the given QgsVertexId belongs to the geometry
- use hasVertex in deleteVertices, if everything is correct, deleteVertices shouldn't fail
- add tests for hasVertex
- optimize deleteVertices with early returns

ViperMiniQ committed:
sipify

ViperMiniQ committed:
fix docstring, explain vertex deletion for invalid vertices

uclaros reviewed (approved):
Great work!
Looks good now 👍
